crates/wasmtime/src/runtime/type_registry.rs

```diff
@@ -809,7 +809,24 @@ impl TypeRegistryInner {
             "type is not canonicalized for runtime usage: {ty:?}"
         );
 
-        let gc_layout = gc_runtime.layouts().gc_layout(&ty.composite_type);
+        assert!(!ty.composite_type.shared);
+        let gc_layout = match &ty.composite_type.inner {
+            wasmtime_environ::WasmCompositeInnerType::Func(_) => None,
+            wasmtime_environ::WasmCompositeInnerType::Array(a) => Some(
+                gc_runtime
+                    .expect("must have a GC runtime to register array types")
+                    .layouts()
+                    .array_layout(a)
+                    .into(),
+            ),
+            wasmtime_environ::WasmCompositeInnerType::Struct(s) => Some(
+                gc_runtime
+                    .expect("must have a GC runtime to register array types")
+                    .layouts()
+                    .struct_layout(s)
+                    .into(),
+            ),
+        };
```

Comment:
I ended up removing the refactor we used in this PR. Though indeed we could have done `gc_runtime.map(...).flatten()` and ended up with the same result, I'm concerned that @fitzgen probably is using these `expect`s as an early warning sign for winch, e.g.